File: golang/the_k_weakest_rows_in_a_matrix.go
package golang

func kWeakestRows(mat [][]int, k int) []int {
	ans := make([]int, k)
	row, col := len(mat), len(mat[0])
	d := make(map[int]int)
	idx := 0

	for x := 0; x < col; x++ {
		for y := 0; y < row; y++ {
			if _, ok := d[y]; !ok {
				if mat[y][x] == 0 {
					d[y] = idx
					idx++
				}
			}
		}
	}

	if len(d) < k {
		for i := 0; i < row; i++ {
			if _, ok := d[i]; !ok {
				d[i] = idx
				idx++
			}
		}
	}

	for key, val := range d {
		if val < k {
			ans[val] = key
		}
	}

	return ans
}

File: python3/the_k_weakest_rows_in_a_matrix.py
from typing import List

"""
Space   : O(m)
Time    : O(mn + m log m)
"""


class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        m, n = len(mat), len(mat[0])
        pairs = []

        for i in range(m):
            sol = 0
            j = 0
            while j < n:
                if mat[i][j] == 1:
                    sol += 1
                else:
                    break
                j += 1

            pairs.append((sol, i))
            
        pairs.sort(key=lambda x: (x[0], x[1]))
        
        return [x[1] for x in pairs[:k]]
